Horizon/Horizon/Sources/Features/Notebook/Course/Model/GetCourseNotesInteractor.swift

```diff
@@ -0,0 +1,100 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2024-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import Combine
+import Foundation
+
+struct NotebookCourseNote {
+    let date: Date
+    let id: String
+    let note: String
+    let type: NotebookNoteLabel?
+}
+
+enum NotebookNoteLabel: String, CaseIterable {
+    case confusing = "Confusing"
+    case important = "Important"
+}
+
+class GetCourseNotesInteractor {
+
+    // MARK: - Dependencies
+
+    let courseNotesRepository: CourseNotesRepository
+
+    // MARK: - Properties
+
+    let publisher = PassthroughSubject<[NotebookCourseNote], Never>()
+
+    private var cancellables = Set<AnyCancellable>()
+
+    // MARK: - Init
+
+    init(courseNotesRepository: CourseNotesRepository) {
+        self.courseNotesRepository = courseNotesRepository
+    }
+
+    // MARK: - Public
+
+    func get() -> AnyPublisher<[NotebookCourseNote], Never> {
+        publisher.eraseToAnyPublisher()
+    }
+
+    func search(courseId: String, text: String = "", filter: NotebookNoteLabel? = nil) {
```

Comment:
Take a look at the `CourseListInteractor.swift` file for an example on how to simplify the reactive stream. You could have one function that retrieves the data and an other one that sets a filter. The function that retrieves the data can return an `AnyPublisher` so the memory handling is up to the caller site not the interactor itself.

Reply:
That's cool, i like that pattern